### PR TITLE
feat: add LDAP authentication module with JWT issuance

### PR DESCRIPTION
**Pull request by mundele2004**
## Pull Request Template

**Description:**

- Added a plug-and-play LDAP authentication module under `internal/auth/ldapauth`.
- The module performs `Bind(username, password)` authentication against an LDAP server and issues a JWT token upon successful login.
- Addresses the LDAP Integration proposal for GoFr Framework Extension as per SOC.
- Closes: issue #2014

**Breaking Changes (if applicable):**

- None. This module is self-contained and does not interfere with any existing authentication mechanisms or routing middleware.

**Additional Information:**

- External Library used: [`github.com/go-ldap/ldap/v3`](https://pkg.go.dev/github.com/go-ldap/ldap/v3)
- JWT is issued using standard library packages (`crypto/rand`, `encoding/base64`) with HMAC SHA256 signing.
- Unit tests include edge cases using a mock LDAP dialer function.
- See `mainFile/main.go` for demo and `testuser.ldif` for testing against a local OpenLDAP instance.
- Sample test user:

### Summary

This PR introduces a standalone LDAP authentication module (`internal/auth/ldapauth`) for the GoFr framework. It uses `Bind(username, password)` to validate users and issues a JWT on successful login.

---

### What's Included

- `ldapauth.go`: Core logic for LDAP authentication and JWT issuance
- `ldapauth_test.go`: Unit tests with mock LDAP dialer
- `main.go`: Demo login flow using `/login` route
- `testuser.ldif`: Sample LDAP user for testing

---

### How to Test

1. Run local LDAP server (e.g., using Docker: osixia/openldap)
2. Load test user using `testuser.ldif`:
   ```bash
   docker cp mainFile/testuser.ldif myldap:/testuser.ldif
   docker exec -it myldap ldapadd -x -D "cn=admin,dc=example,dc=com" -w admin -f /testuser.ldif
3. Run the server:- go run mainFile/main.go
4. Test login:
curl -X POST http://localhost:8080/login \
  -H "Content-Type: application/json" \
  -d '{"username":"testuser", "password":"testpass"}'

**Checklist:**

- [x] I have formatted my code using `goimport` and `golangci-lint`.
- [x] All new code is covered by unit tests.
- [x] This PR does not decrease the overall code coverage.
- [x] I have reviewed the code comments and documentation for clarity.

### CODE REVIEW

File: internal/auth/ldapauth/ldapauth.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    "github.com/go-ldap/ldap/v3"
+    "github.com/golang-jwt/jwt/v4"
+)
+
+type Config struct {
+    Addr         string
+    BaseDN       string
+    BindUserDN   string // optional
+    BindPassword string // optional
+    JWTSecret    string
```
Comment (gizmo-rt) [inline internal/auth/ldapauth/ldapauth.go]: It would be helpful if we could have comments on usage of these fields

File: internal/auth/ldapauth/ldapauth.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+func New(cfg Config) *Authenticator {
+    return &Authenticator{
+        cfg: cfg,
+        dialFn: ldap.DialURL,
+    }
+}
+
+func (a *Authenticator) Authenticate(username, password string) (string, error) {
+    // l, err := ldap.DialURL("ldap://" + a.cfg.Addr)
+    l, err := a.dialFn("ldap://" + a.cfg.Addr)
```
Comment (gizmo-rt) [inline internal/auth/ldapauth/ldapauth.go]: Could you elaborate on what `l` is ?

Comment (gizmo-rt) [reply internal/auth/ldapauth/ldapauth.go]: Also, was wondering what possible validations we could have for `a.cfg.Addr` ?

File: internal/auth/ldapauth/ldapauth.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    if err != nil {
+        return "", fmt.Errorf("failed to connect LDAP: %w", err)
+    }
+    defer l.Close()
+
+    // err = l.StartTLS(&tls.Config{InsecureSkipVerify: true})
+    // if err != nil {
+    //     return "", fmt.Errorf("TLS error: %w", err)
+    // }
```
Comment (gizmo-rt) [inline internal/auth/ldapauth/ldapauth.go]: If the code is unreachable then its ideal to remove unless there is a specific need for the same

File: internal/auth/ldapauth/ldapauth.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+
+type Authenticator struct {
+    cfg Config
+    dialFn dialerFunc
+}
+
+func New(cfg Config) *Authenticator {
+    return &Authenticator{
+        cfg: cfg,
```
Comment (gizmo-rt) [inline internal/auth/ldapauth/ldapauth.go]: Is the config specific to the LDAP library being used ?

File: internal/auth/ldapauth/ldapauth.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    if err := l.Bind(userDN, password); err != nil {
+        return "", fmt.Errorf("invalid credentials")
+    }
+
+    token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
+        "sub": username,
+        "exp": time.Now().Add(1 * time.Hour).Unix(),
+    })
+
+    signed, err := token.SignedString([]byte(a.cfg.JWTSecret))
```
Comment (gizmo-rt) [inline internal/auth/ldapauth/ldapauth.go]: This should preferably be decomposed into smaller, testable sections

File: internal/auth/ldapauth/ldapauth.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
+package ldapauth
```
Comment (gizmo-rt) [inline internal/auth/ldapauth/ldapauth.go:1]: For package names you can follow https://go.dev/doc/effective_go#package-names

File: internal/auth/ldapauth/ldapauth.go (hunk 1 above)
Comment (mundele2004) [reply internal/auth/ldapauth/ldapauth.go]: Addr         string  // LDAP server address (e.g., "localhost:389" or "ldap.example.com:389")
    BaseDN       string  // Base distinguished name (DN) used to search for users (e.g., "dc=example,dc=com")
    BindUserDN   string // Optional: DN of a service account to perform search operations.
                                   // Required if anonymous search is disabled on the LDAP server.
BindPassword string // Optional: Password for the service account specified in BindUserDN.
                        // Required only if BindUserDN is set.
    JWTSecret    string // Secret key used to sign JWT tokens issued after successful authentication.                                            

**like this**

File: internal/auth/ldapauth/ldapauth.go (hunk 2 above)
Comment (mundele2004) [reply internal/auth/ldapauth/ldapauth.go]: l is the LDAP connection (*ldap.Conn) returned by the dialer.

We should validate a.cfg.Addr for non-empty and proper host:port format using net.SplitHostPort.

File: internal/auth/ldapauth/ldapauth.go (hunk 3 above)
Comment (mundele2004) [reply internal/auth/ldapauth/ldapauth.go]: Removed the commented StartTLS block as it is currently unused.

File: internal/auth/ldapauth/ldapauth.go (hunk 4 above)
Comment (mundele2004) [reply internal/auth/ldapauth/ldapauth.go]: The Config struct here is not directly from the go-ldap/ldap/v3 library — it's a custom wrapper defined within this module to keep the Authenticator decoupled from external configs.

This design allows flexibility — if we change or extend the backend (e.g. add StartTLS, change port or baseDN logic), we just update our config struct without impacting the rest of the app.

// Config holds LDAP server connection details and base search parameters.
// It is application-defined and not tied to the go-ldap library.

 ```
 type Config struct {
       Addr     string
       BaseDN   string
       BindDN   string
       BindPass string
}
```

**Comment by mundele2004** [comment]
Hi @gizmo-rt 👋
✅ All edge cases tested and working:
- Valid login
- Invalid password
- Unknown user
- LDAP unreachable

Updated main.go + test cases as requested.


@gizmo-rt I’ve added the requested edge case tests. Let me know if any further changes are needed. ✅


✅ Added all suggested edge case tests including:
- LDAP server unreachable
- Unauthorized bind
- Invalid credentials
- Expired password simulation
- DN found but user info retrieval failed
- JWT generation failure
- Expired JWT handling

🧪 All tests pass locally.


PR is ready for review. Thanks!

**Comment by mundele2004** [comment]
@Umang01-hash @coolwednesday  Is there any way or command I could use to run Code Quality checks on my local machine in the same way they are executed in the pipeline?

**Comment by Umang01-hash** [comment]
> @Umang01-hash @coolwednesday Is there any way or command I could use to run Code Quality checks on my local machine in the same way they are executed in the pipeline?

Hey @mundele2004! To run these code qaulity checks you can run `golangci-lint run` command on your local and see. Just make sure the linter version used in pipeline and the one installed on your system is same.

**Comment by mundele2004** [comment]
✅ Resolved all `golangci-lint` issues and force-pushed updated changes.
All tests are passing locally. Awaiting review. 🚀
plz check @Umang01-hash

File: internal/auth/ldapauth/ldapauth.go [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+	BindUserDN string // Optional: DN of a service account to perform search operations.
+
+	// Required if anonymous search is disabled on the LDAP server.
+
+	BindPassword string // Optional: Password for the service account specified in BindUserDN.
+
+	// Required only if BindUserDN is set.
+
+	JWTSecret string // Secret key used to sign JWT tokens issued after successful authentication.
```
Comment (akshat-kumar-singhal) [inline internal/auth/ldapauth/ldapauth.go]: Issuance of JWT should be handled by an independent package as JWT may be issued based on LDAP, API Key/Secret or any other authentication mechanism.

Comment (akshat-kumar-singhal) [reply internal/auth/ldapauth/ldapauth.go]: So it's better to have an interface (say JWTIssuer) in the config which can be used to issue the tokens.

File: internal/auth/ldapauth/ldapauth.go [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+	result, err := conn.Search(searchReq)
+
+	if err != nil {
+
+		return "", fmt.Errorf("search error: %w", err)
+
+	}
+
+	if len(result.Entries) != 1 {
```
Comment (akshat-kumar-singhal) [inline internal/auth/ldapauth/ldapauth.go]: User not found and multiple users found should be handled separately. From an HTTP standpoint, user not found would result in 401 Response Code, whereas multiple users found would typically be a 500 scenario - because the user didn't enter incorrect credentials, but credentials store is misconfigured and the "admin" should be getting notified to fix it.

File: internal/auth/ldapauth/ldapauth.go [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+
+		return "", err
+
+	}
+
+	return token, nil
+
+}
+
+func (a *Authenticator) lookupUserDN(conn Conn, username string) (string, error) {
```
Comment (akshat-kumar-singhal) [inline internal/auth/ldapauth/ldapauth.go]: This should return other information associated with the user in the LDAP Directory as it may be needed to be embedded into the token/response/session

**Comment by mundele2004** [comment]
Final PR after cleanup and linting. Ready for review 🚀
cc @akshat-kumar-singhal @Umang01-hash

**Comment by mundele2004** [comment]
✅ Resolved all golangci-lint issues .
All tests are passing locally. . Ready for review 🚀
 @akshat-kumar-singhal  @Umang01-hash

File: internal/auth/ldapauth/ldapauth.go (hunk 7 above)
Comment (akshat-kumar-singhal) [reply internal/auth/ldapauth/ldapauth.go]: @mundele2004 this is not resolved

**Comment by akshat-kumar-singhal** [comment]
> ✅ Resolved all golangci-lint issues . All tests are passing locally. . Ready for review 🚀 @akshat-kumar-singhal @Umang01-hash

@mundele2004 Please address all the feedback/comments on the PR - don't mark them resolved without a response/change.

File: internal/auth/ldapauth/ldapauth.go (hunk 7 above)
Comment (mundele2004) [reply internal/auth/ldapauth/ldapauth.go]: Done Plz check it @akshat-kumar-singhal

**Comment by mundele2004** [comment]
Please check — I have fixed the linter issues such as whitespace, cuddled statements, etc.
You also check and let me know. If there are still 2 `gci` issues, please tell me how I can solve them, because I have tried a lot but am unable to fix them. Kindly guide me.

**$ golangci-lint run
internal\auth\ldapauth\ldapauth.go:7:1: File is not properly formatted (gci)
        "github.com/go-ldap/ldap/v3"
^
internal\auth\ldapauth\ldapauth_test.go:7:1: File is not properly formatted (gci)
        "github.com/go-ldap/ldap/v3"
^
2 issues:
* gci: 2** 

@akshat-kumar-singhal @Umang01-hash

File: internal/auth/ldapauth/ldapauth.go (hunk 7 above)
Comment (mundele2004) [reply internal/auth/ldapauth/ldapauth.go]: Please check — I have fixed the linter issues such as whitespace, cuddled statements, etc.
You also check and let me know. If there are still 2 gci issues, please tell me how I can solve them, because I have tried a lot but am unable to fix them. Kindly guide me.

**$ golangci-lint run
internal\auth\ldapauth\ldapauth.go:7:1: File is not properly formatted (gci)
"github.com/go-ldap/ldap/v3"
^
internal\auth\ldapauth\ldapauth_test.go:7:1: File is not properly formatted (gci)
"github.com/go-ldap/ldap/v3"
^
2 issues:

gci: 2**
@akshat-kumar-singhal @Umang01-hash

**Comment by Umang01-hash** [comment]
> Please check — I have fixed the linter issues such as whitespace, cuddled statements, etc. You also check and let me know. If there are still 2 `gci` issues, please tell me how I can solve them, because I have tried a lot but am unable to fix them. Kindly guide me.
> 
> **$ golangci-lint run internal\auth\ldapauth\ldapauth.go:7:1: File is not properly formatted (gci) "github.com/go-ldap/ldap/v3" ^ internal\auth\ldapauth\ldapauth_test.go:7:1: File is not properly formatted (gci) "github.com/go-ldap/ldap/v3" ^ 2 issues:
> 
> * gci: 2**
> 
> @akshat-kumar-singhal @Umang01-hash

@mundele2004 You can fix gci import order issues by installing and running the gci tool.
```
go install github.com/daixiang0/gci@latest
```

and then use `gci write .` inside the directory **ldapauth**. It should fix these errors.

**Comment by mundele2004** [comment]
> > Please check — I have fixed the linter issues such as whitespace, cuddled statements, etc. You also check and let me know. If there are still 2 `gci` issues, please tell me how I can solve them, because I have tried a lot but am unable to fix them. Kindly guide me.
> > **$ golangci-lint run internal\auth\ldapauth\ldapauth.go:7:1: File is not properly formatted (gci) "github.com/go-ldap/ldap/v3" ^ internal\auth\ldapauth\ldapauth_test.go:7:1: File is not properly formatted (gci) "github.com/go-ldap/ldap/v3" ^ 2 issues:
> > 
> > * gci: 2**
> > 
> > @akshat-kumar-singhal @Umang01-hash
> 
> @mundele2004 You can fix gci import order issues by installing and running the gci tool.
> 
> ```
> go install github.com/daixiang0/gci@latest
> ```
> 
> and then use `gci write .` inside the directory **ldapauth**. It should fix these errors.


@Umang01-hash I tried the steps you suggested, but those 2 issues are still not resolved. Please guide me on how to fix those errors.

**Comment by mundele2004** [comment]
> > > Please check — I have fixed the linter issues such as whitespace, cuddled statements, etc. You also check and let me know. If there are still 2 `gci` issues, please tell me how I can solve them, because I have tried a lot but am unable to fix them. Kindly guide me.
> > > **$ golangci-lint run internal\auth\ldapauth\ldapauth.go:7:1: File is not properly formatted (gci) "github.com/go-ldap/ldap/v3" ^ internal\auth\ldapauth\ldapauth_test.go:7:1: File is not properly formatted (gci) "github.com/go-ldap/ldap/v3" ^ 2 issues:
> > > 
> > > * gci: 2**
> > > 
> > > @akshat-kumar-singhal @Umang01-hash
> > 
> > 
> > @mundele2004 You can fix gci import order issues by installing and running the gci tool.
> > ```
> > go install github.com/daixiang0/gci@latest
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and then use `gci write .` inside the directory **ldapauth**. It should fix these errors.
> 
> @Umang01-hash I tried the steps you suggested, but those 2 issues are still not resolved. Please guide me on how to fix those errors.

@Umang01-hash 
I have resolved all the issues, including the merge conflicts and the gci import order problems. I installed the gci tool and ran gci write . inside the ldapauth directory as suggested, and now golangci-lint passes without errors.

Everything is working fine now. ✅

**Comment by mundele2004** [comment]
Please tell me what I should do to fix this:
**Workflow-Pipeline / PKG Unit Testing (v1.22)🛠 (pull\_request)**
@Umang01-hash

**Comment by mundele2004** [comment]
Please tell me what I should do to fix this:
Workflow-Pipeline / PKG Unit Testing (v1.22)🛠 (pull_request)
@Umang01-hash  @akshat-kumar-singhal

**Comment by mundele2004** [comment]
Please tell me what I should do to fix this:
Workflow-Pipeline / PKG Unit Testing (v1.22)🛠 (pull_request)
@Umang01-hash  @akshat-kumar-singhal

**Comment by Umang01-hash** [comment]
@mundele2004 Thanks for your contribution to GoFr.
As this PR has been inactive for quite some time, and considering the nature of the file changes, I believe it would be more effective to first align on the approach and implementation. I'm closing this PR for now.

If you're still interested in adding LDAP authentication to GoFr, let's start fresh with a new implementation. Please ensure it aligns with our existing module structure and maintains the ease of integration for users.
